// tiktok-live-connector/dist/lib/webcastDataConverter.js
"use strict";

/**
 * This ugly function brings the nested protobuf objects to a flat level
 * In addition, attributes in "Long" format are converted to strings (e.g. UserIds)
 * This makes it easier to handle the data later, since some libraries have problems to serialize this protobuf specific data.
 */
function simplifyObject(webcastObject) {
  if (webcastObject.questionDetails) {
    Object.assign(webcastObject, webcastObject.questionDetails);
    delete webcastObject.questionDetails;
  }

  if (webcastObject.user) {
    Object.assign(webcastObject, getUserAttributes(webcastObject.user));
    delete webcastObject.user;
  }

  if (webcastObject.event) {
    Object.assign(webcastObject, webcastObject.event);
    delete webcastObject.event;
  }

  if (webcastObject.eventDetails) {
    Object.assign(webcastObject, webcastObject.eventDetails);
    delete webcastObject.eventDetails;
  }

  if (webcastObject.battleUsers) {
    let battleUsers = [];
    webcastObject.battleUsers.forEach(user => {
      var _user$battleGroup;

      if (user !== null && user !== void 0 && (_user$battleGroup = user.battleGroup) !== null && _user$battleGroup !== void 0 && _user$battleGroup.user) {
        battleUsers.push(getUserAttributes(user.battleGroup.user));
      }
    });
    webcastObject.battleUsers = battleUsers;
  }

  if (webcastObject.battleItems) {
    webcastObject.battleArmies = [];
    webcastObject.battleItems.forEach(battleItem => {
      battleItem.battleGroups.forEach(battleGroup => {
        let group = {
          hostUserId: battleItem.hostUserId.toString(),
          points: parseInt(battleGroup.points),
          participants: []
        };
        battleGroup.users.forEach(user => {
          group.participants.push(getUserAttributes(user));
        });
        webcastObject.battleArmies.push(group);
      });
    });
    delete webcastObject.battleItems;
  }

  if (webcastObject.giftId) {
    var _webcastObject$giftDe;

    // Convert to boolean
    webcastObject.repeatEnd = !!webcastObject.repeatEnd; // Add previously used JSON structure (for compatibility reasons)
    // Can be removed soon

    webcastObject.gift = {
      gift_id: webcastObject.giftId,
      repeat_count: webcastObject.repeatCount,
      repeat_end: webcastObject.repeatEnd ? 1 : 0,
      gift_type: (_webcastObject$giftDe = webcastObject.giftDetails) === null || _webcastObject$giftDe === void 0 ? void 0 : _webcastObject$giftDe.giftType
    };

    if (webcastObject.giftDetails) {
      Object.assign(webcastObject, webcastObject.giftDetails);
      delete webcastObject.giftDetails;
    }

    if (webcastObject.giftImage) {
      Object.assign(webcastObject, webcastObject.giftImage);
      delete webcastObject.giftImage;
    }

    if (webcastObject.giftExtra) {
      Object.assign(webcastObject, webcastObject.giftExtra);
      delete webcastObject.giftExtra;

      if (webcastObject.receiverUserId) {
        webcastObject.receiverUserId = webcastObject.receiverUserId.toString();
      }

      if (webcastObject.timestamp) {
        webcastObject.timestamp = parseInt(webcastObject.timestamp);
      }
    }
  }

  if (webcastObject.emote) {
    var _webcastObject$emote, _webcastObject$emote2, _webcastObject$emote3;

    webcastObject.emoteId = (_webcastObject$emote = webcastObject.emote) === null || _webcastObject$emote === void 0 ? void 0 : _webcastObject$emote.emoteId;
    webcastObject.emoteImageUrl = (_webcastObject$emote2 = webcastObject.emote) === null || _webcastObject$emote2 === void 0 ? void 0 : (_webcastObject$emote3 = _webcastObject$emote2.image) === null || _webcastObject$emote3 === void 0 ? void 0 : _webcastObject$emote3.imageUrl;
    delete webcastObject.emote;
  }

  if (webcastObject.treasureBoxUser) {
    var _webcastObject$treasu, _webcastObject$treasu2, _webcastObject$treasu3, _webcastObject$treasu4;

    // holy crap
    Object.assign(webcastObject, getUserAttributes(((_webcastObject$treasu = webcastObject.treasureBoxUser) === null || _webcastObject$treasu === void 0 ? void 0 : (_webcastObject$treasu2 = _webcastObject$treasu.user2) === null || _webcastObject$treasu2 === void 0 ? void 0 : (_webcastObject$treasu3 = _webcastObject$treasu2.user3[0]) === null || _webcastObject$treasu3 === void 0 ? void 0 : (_webcastObject$treasu4 = _webcastObject$treasu3.user4) === null || _webcastObject$treasu4 === void 0 ? void 0 : _webcastObject$treasu4.user) || {}));
    delete webcastObject.treasureBoxUser;
  }

  if (webcastObject.treasureBoxData) {
    Object.assign(webcastObject, webcastObject.treasureBoxData);
    delete webcastObject.treasureBoxData;
    webcastObject.timestamp = parseInt(webcastObject.timestamp);
  }

  return Object.assign({}, webcastObject);
}

function getUserAttributes(webcastUser) {
  var _webcastUser$userId, _webcastUser$profileP, _webcastUser$extraAtt, _userAttributes$userB, _userAttributes$userB2, _userAttributes$userB3;

  let userAttributes = {
    userId: (_webcastUser$userId = webcastUser.userId) === null || _webcastUser$userId === void 0 ? void 0 : _webcastUser$userId.toString(),
    uniqueId: webcastUser.uniqueId !== '' ? webcastUser.uniqueId : undefined,
    nickname: webcastUser.nickname !== '' ? webcastUser.nickname : undefined,
    profilePictureUrl: (_webcastUser$profileP = webcastUser.profilePicture) === null || _webcastUser$profileP === void 0 ? void 0 : _webcastUser$profileP.urls[2],
    followRole: (_webcastUser$extraAtt = webcastUser.extraAttributes) === null || _webcastUser$extraAtt === void 0 ? void 0 : _webcastUser$extraAtt.followRole,
    userBadges: mapBadges(webcastUser.badges)
  };
  userAttributes.isModerator = userAttributes.userBadges.some(x => x.type && x.type.toLowerCase().includes('moderator'));
  userAttributes.isNewGifter = userAttributes.userBadges.some(x => x.type && x.type.toLowerCase().includes('live_ng_'));
  userAttributes.isSubscriber = userAttributes.userBadges.some(x => x.url && x.url.toLowerCase().includes('/sub_'));
  userAttributes.topGifterRank = (_userAttributes$userB = (_userAttributes$userB2 = userAttributes.userBadges.find(x => x.url && x.url.includes('/ranklist_top_gifter_'))) === null || _userAttributes$userB2 === void 0 ? void 0 : (_userAttributes$userB3 = _userAttributes$userB2.url.match(/(?<=ranklist_top_gifter_)(\d+)(?=.png)/g)) === null || _userAttributes$userB3 === void 0 ? void 0 : _userAttributes$userB3.map(Number)[0]) !== null && _userAttributes$userB !== void 0 ? _userAttributes$userB : null;
  return userAttributes;
}

function mapBadges(badges) {
  let simplifiedBadges = [];

  if (Array.isArray(badges)) {
    badges.forEach(innerBadges => {
      if (Array.isArray(innerBadges.badges)) {
        innerBadges.badges.forEach(badge => {
          simplifiedBadges.push(Object.assign({}, badge));
        });
      }

      if (Array.isArray(innerBadges.imageBadges)) {
        innerBadges.imageBadges.forEach(badge => {
          if (badge && badge.image && badge.image.url) {
            simplifiedBadges.push({
              type: 'image',
              displayType: badge.displayType,
              url: badge.image.url
            });
          }
        });
      }
    });
  }

  return simplifiedBadges;
}

module.exports = {
  simplifyObject
};
